pkg/processor/runtime/golang/runtime.go

```diff
@@ -119,17 +151,58 @@ func (g *golang) callEntrypoint(event nuclio.Event, functionLogger logger.Logger
 		}
 	}()
 
-	// before we call, save timestamp
-	startTime := time.Now()
+	ctx, cancel := context.WithCancel(context.Background())
+	g.cancelEventHandlingChan <- cancel
+	responseChan := make(chan *processingResponse, 1)
+	// Run the function in a goroutine
+	go func() {
+		defer func() {
+			if err := recover(); err != nil {
+				callStack := debug.Stack()
+				functionLogger.WarnWith("Panic caught in event handler",
+					"err",
+					err,
+					"stack",
+					string(callStack))
+			}
+		}()
+		// before we call, save timestamp
+		startTime := time.Now()
+
+		// call entrypoint
+		response, responseErr = g.entrypoint(g.Context, event)
 
-	response, responseErr = g.entrypoint(g.Context, event)
+		// calculate how long it took to invoke the function
+		callDuration := time.Since(startTime)
 
-	// calculate how long it took to invoke the function
-	callDuration := time.Since(startTime)
+		// signal that results
+		responseChan <- &processingResponse{
+			response,
+			responseErr,
+		}
 
-	// add duration to sum
-	g.Statistics.DurationMilliSecondsSum += uint64(callDuration.Nanoseconds() / 1000000)
-	g.Statistics.DurationMilliSecondsCount++
+		// add duration to sum
+		g.Statistics.DurationMilliSecondsSum += uint64(callDuration.Nanoseconds() / 1000000)
+		g.Statistics.DurationMilliSecondsCount++
+	}()
+	select {
+	case result := <-responseChan:
+		select {
+		case cancelEventHandling := <-g.cancelEventHandlingChan:
+			// cancelling cancel-context
+			defer cancelEventHandling()
+		default:
+		}
+		response = result.response
+		responseErr = result.responseErr
```

Comment:
`response` and `responseErr` are already set in the goroutine (line 173).
If you need to signal that the goroutine is done, you can use a simpler `struct{}{}` channel.

Reply:
@TomerShor the reason why I did it like that is since this code is super critical, I'd prefer to be on the safest possible side. But I don't mind either way

Reply:
@TomerShor actually, after adding sleep to the test, it caught data race with this approach (with struct). So I stick with the original one.